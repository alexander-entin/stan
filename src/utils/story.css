.diff {
	/* customization for the top-level <div> wrapper that contains the diff */
	font-family: ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
	font-size: 1em;
	white-space: pre;
}

.deletion {
	/* customization for deleted lines (<div>) */
	color: #ff0000;
}

.addition {
	/* customization for added lines (<div>) */
	color: #00dd00;
}

.unchanged {
	/* customization for unchanged lines (<div>) */
	color: gray;
}
