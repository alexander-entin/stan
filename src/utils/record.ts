import { $global, $event, snapshot, subscribe } from '../utils/stan'

if (typeof window !== 'undefined') {
	setTimeout(() => {
		let initial
		let events = []
		window.record = () => {
			if (events.length) {
				console.log({
					initial,
					events,
				})
			}
			initial = snapshot($global)
		}
		record()
		subscribe($event, () => {
			events.push(snapshot($event))
		})
	}, 1)
}
