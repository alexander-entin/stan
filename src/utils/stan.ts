import { proxy, subscribe, snapshot, useSnapshot, ref } from 'valtio'
import { subscribeKey, derive, watch, devtools } from 'valtio/utils'
import { nanoid } from 'nanoid'
import { path } from 'rambda'

export { subscribe, subscribeKey, snapshot, watch, derive, ref }
export const useStan = useSnapshot

export type Entity = {
	id: string,
	[key: string]: unknown,
}

export type SyncState = {
	status?: string,
	error?: unknown,
	pull: {
		[reqId: string]: {
			query: unknown,
			at: number,
			doneAt?: number,
			error?: unknown,
		},
	},
	push: {
		[reqId: string]: {
			method: string,
			body: Entity,
			prev: Entity,
			at: number,
			doneAt?: number,
			error?: unknown,
		},
	}
}

export const $global = proxy({
	sync: {} as Record<string, SyncState>,
})
export const $sync = $global.sync
export const $event = proxy({
	id: nanoid(),
	at: Date.now(),
	type: 'init',
	data: [] as any[],
})

type Reactor = (...payload: any[]) => void
const reactors = {} as { [event: string]: { [module: string]: Reactor } }
export function createReactor<F extends Reactor>(module: string, event: string, fn: F) {
	let rs = reactors[event] ?? {}
	reactors[event] = rs
	let id = module ?? Object.keys(rs).length
	rs[id] = fn
	return dispatch({ type: event }) as F
}

export type EventMeta = {
	type: string,
	id: string,
	at: number,
}
export type Event = EventMeta & { data?: any[] }

let initial
export const dispatch = ({ id, at, type }: Partial<EventMeta>) => (...payload: unknown[]) => {
	if (import.meta.env.DEV && typeof window === 'object') {
		initial ||= snapshot($global)
	}
	$event.id = id ?? nanoid()
	$event.at = at ?? Date.now()
	$event.type = type!
	$event.data = payload
	console.warn(type, JSON.stringify(payload))
	Object.values(reactors[type!]).forEach(f => {
		// console.warn(type, JSON.stringify(payload), { f })
		f(...payload)
	})
	// console.warn(snapshot($global))
}

function isVirtualProp(o, k) {
	let d = Object.getOwnPropertyDescriptor(o, k) || {}
	return d.get || d.set
}

function isDataProp(o, k) {
	return !isVirtualProp(o, k) && typeof o[k] !== 'function'
}

export function createStan<T extends object>(name: string, stan: T): T {
	Object.keys(stan).forEach(k => {
		if (!isVirtualProp(stan, k)
			&& typeof stan[k] === 'function'
			&& k.startsWith('on')
			&& k[2] === k[2]?.toUpperCase()
		) {
			let event = `${name}.${k}`
			stan[k] = createReactor(name, event, stan[k])
		}
	})
	$global[name] = stan
	return $global[name] as T
}

function dive(o, f, p = [] as string[]) {
	Object.keys(o).forEach(k => {
		if (!isDataProp(o, k)) return
		f(o, k, p)
		if (typeof o[k] === 'object') {
			dive(o[k], f, [...p, k])
		}
	})
}

export function reset($stan, snap = initial) {
	if (!snap) return
	dive(snap, (s, k, p) => {
		const $ = path(p, $stan)
		if (!isDataProp($, k)) return
		if (typeof s[k] !== 'object' || typeof $[k] !== 'object') {
			$[k] = structuredClone(s[k])
		}
	})
	dive($stan, ($, k, p) => {
		const s = path(p, snap)
		if (!(k in s)) {
			delete $[k]
		}
	})
}

export function integrated() {
	return import.meta.env.MODE !== 'test' && import.meta.env.MODE !== 'story'
}

export type Paths = string | string[] | string[][]
export type SubscribeCallback = (ops: unknown[][]) => void
export function subscribePaths(proxy: object, paths: Paths, fn: SubscribeCallback, ...rest) {
	if (typeof paths === 'string') paths = paths.split(',')
	paths = paths.map(x => typeof x === 'string' ? x.split('.') : x)
	return subscribe(proxy, ops => {
		ops = ops.filter(([_, path]) =>
			(paths as string[][]).find(p => p.every((x, i) => path[i] === x))
		)
		ops.length && fn(ops)
	}, ...rest)
}

if (import.meta.env.DEV && typeof window === 'object') {
	Object.assign(window, { $global, $event, proxy, subscribe, watch, derive, snapshot, reset })
	devtools($global, { name: '$global', enabled: true })

	let snap
	let events = [] as Event[]
	function story() {
		const indent = snap ? '\t' : ''
		let json = ['[', ...events.map(({ type, data = [] }) => {
			let rest = [data]
			if (data.length === 0) {
				rest = []
			} else if (data.length === 1 && !Array.isArray(data[0])) {
				rest = data
			}
			return `\t${JSON.stringify([type, ...rest])},`
		}), ']'].join(`\n${indent}`)
		if (snap) {
			let init = JSON.stringify(snap, null, '\t')
			init = init.split('\n').map((r, i) => `${i ? '\t' : ''}${r}`).join('\n')
			json = `{\n\tinitial: ${init},\n\tevents: ${json}\n}`
		}
		const val = snap ? { initial: snap, events } : events
		events = []
		snap = snapshot($global)
		dive(snap, (s, k, p) => {
			const $ = path(p, $global)
			if (!isDataProp($, k)) {
				delete s[k]
			}
		})
		console.log(json)
		return val
	}
	subscribe($event, () => {
		events.push(snapshot($event) as Event)
	})
	window['story'] = story
}
