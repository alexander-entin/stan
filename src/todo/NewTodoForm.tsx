import { useStan } from '../utils/stan'
import { $todo } from './$todo'

export function NewTodoForm() {
	console.log('NewTodoForm.render')
	const { draft, onDraft, onAdd } = useStan($todo)
	return (
		<form
			onSubmit={e => {
				e.preventDefault()
				onAdd()
			}}
		>
			<input
				type="text"
				placeholder="New Todo"
				value={draft}
				onInput={e => onDraft(e.currentTarget.value)}
			/>
			<button
				className="btn"
				type="submit"
				// disabled={!draft.trim()}
			>
				Create
			</button>
		</form>
	)
}