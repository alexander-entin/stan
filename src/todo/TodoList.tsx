import { useStan } from '../utils/stan'
import { $todo } from './$todo'
import { Todo } from './Todo'

export function TodoList() {
	const { filtered } = useStan($todo)
	console.log('TodoList.render', filtered)
	return (
		<ul>
			{filtered.map(x => (
				<Todo id={x.id} key={x.id} />
			))}
		</ul>
	)
}
