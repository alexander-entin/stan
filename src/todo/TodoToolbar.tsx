import { useStan } from '../utils/stan'
import { $todo } from './$todo'

export function TodoToolbar() {
	// console.log('TodoToolbar.render')
	const { map, filter, onFilter, onDeleteDone } = useStan($todo)
	return <>
		<input
			type="checkbox"
			checked={filter === '!done'}
			onChange={() => onFilter(filter === '!done' ? '' : '!done')}
		/>
		Hide completed
		<button
			className='btn'
			onClick={() => onDeleteDone()}
			disabled={!Object.keys(map).some(k => map[k].done)}
		>
			Remove completed
		</button>
	</>
}

