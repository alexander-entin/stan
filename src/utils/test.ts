import { it } from 'vitest'

import { dispatch, $global, snapshot } from '../utils/stan'

export function test($stan, cases) {
	Object.entries(cases).forEach(([name, spec]) => {
		it(name, ({ expect }) => {
			spec.events.forEach(({ data, ...meta }) =>
				dispatch(meta)(...(Array.isArray(data) ? data : [data]))
			)
			const snap = snapshot($global)
			expect(snap).toMatchSnapshot()
		})
	})
}