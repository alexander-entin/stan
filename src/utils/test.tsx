import { describe, it } from 'vitest'
import { diff } from 'deep-object-diff'
import deepRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'

import { dispatch, $global, snapshot } from './stan'

export function test(cases, defaultComponents) {
	Object.entries(cases).forEach(([name, { events, components }]) => {
		components = components || defaultComponents
		describe(name, () => {
			let t = new Date('2020-02-02').valueOf()
			let snap = snapshot($global)
			events.forEach((event, idx) => {
				const { id = `#${t}`, at = t, type, data } = event
				dispatch({ id, at, type })(...(Array.isArray(data) ? data : [data]))
				const curr = snapshot($global)
				const d = diff(snap, curr)
				delete d.event
				snap = curr
				it(`${idx}. ${type} > $stan changes`, ({ expect }) => {
					expect(d).toMatchSnapshot()
				})
				const cs = typeof components === 'function' ? components(event, idx) : components
				if (cs) {
					Object.entries(cs).forEach(([name, Component]) => {
						const component = <Component />
						const shallowRenderer = new ShallowRenderer()
						const shallow = shallowRenderer.render(component)
						it(`${idx}. ${type} > ${name} (shallow)`, ({ expect }) => {
							expect(shallow).toMatchSnapshot()
						})
						const deep = deepRenderer.create(component).toJSON()
						it(`${idx}. ${type} > ${name} (deep)`, ({ expect }) => {
							expect(deep).toMatchSnapshot()
						})
					})
				}
				t++
			})
		})
	})
}