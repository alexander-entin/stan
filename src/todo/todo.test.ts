import { test } from '../utils/test'
import Page from './+Page'
import { NewTodoForm } from './NewTodoForm'
import { TodoToolbar } from './TodoToolbar'
import { TodoList } from './TodoList'

const cases = {
	'todo creation': {
		events: [
			{ type: 'todo.onDraft', data: 'buy milk' },
			{ type: 'todo.onAdd' },
		]
	}
}

test(cases, { Page, NewTodoForm, TodoToolbar, TodoList })