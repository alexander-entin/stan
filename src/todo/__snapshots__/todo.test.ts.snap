// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`todo creation > 0. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy milk",
  },
}
`;

exports[`todo creation > 0. todo.onDraft > NewTodoForm (deep) 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onInput={[Function]}
    placeholder="New Todo"
    type="text"
    value="buy milk"
  />
  <button
    className="btn"
    type="submit"
  >
    Create
  </button>
</form>
`;

exports[`todo creation > 0. todo.onDraft > NewTodoForm (shallow) 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onInput={[Function]}
    placeholder="New Todo"
    type="text"
    value="buy milk"
  />
  <button
    className="btn"
    type="submit"
  >
    Create
  </button>
</form>
`;

exports[`todo creation > 0. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy milk"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`todo creation > 0. todo.onDraft > Page (shallow) 1`] = `
<React.Fragment>
  <h1>
    To-do List
  </h1>
  <NewTodoForm />
  <ClientOnly>
    <TodoToolbar />
    <TodoList />
  </ClientOnly>
</React.Fragment>
`;

exports[`todo creation > 0. todo.onDraft > TodoList (deep) 1`] = `<ul />`;

exports[`todo creation > 0. todo.onDraft > TodoList (shallow) 1`] = `<ul />`;

exports[`todo creation > 0. todo.onDraft > TodoToolbar (deep) 1`] = `
[
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
]
`;

exports[`todo creation > 0. todo.onDraft > TodoToolbar (shallow) 1`] = `
<React.Fragment>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  Hide completed
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>
</React.Fragment>
`;

exports[`todo creation > 1. todo.onAdd > $stan changes 1`] = `
{
  "todo": {
    "draft": "",
    "filtered": {
      "0": {
        "id": "#1580601600001",
        "text": "buy milk",
      },
    },
    "map": {
      "#1580601600001": {
        "id": "#1580601600001",
        "text": "buy milk",
      },
    },
  },
}
`;

exports[`todo creation > 1. todo.onAdd > NewTodoForm (deep) 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onInput={[Function]}
    placeholder="New Todo"
    type="text"
    value=""
  />
  <button
    className="btn"
    type="submit"
  >
    Create
  </button>
</form>
`;

exports[`todo creation > 1. todo.onAdd > NewTodoForm (shallow) 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onInput={[Function]}
    placeholder="New Todo"
    type="text"
    value=""
  />
  <button
    className="btn"
    type="submit"
  >
    Create
  </button>
</form>
`;

exports[`todo creation > 1. todo.onAdd > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`todo creation > 1. todo.onAdd > Page (shallow) 1`] = `
<React.Fragment>
  <h1>
    To-do List
  </h1>
  <NewTodoForm />
  <ClientOnly>
    <TodoToolbar />
    <TodoList />
  </ClientOnly>
</React.Fragment>
`;

exports[`todo creation > 1. todo.onAdd > TodoList (deep) 1`] = `
<ul>
  <li
    className=""
  >
    <input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
    buy milk
     
    <button
      onClick={[Function]}
    >
      ☠
    </button>
  </li>
</ul>
`;

exports[`todo creation > 1. todo.onAdd > TodoList (shallow) 1`] = `
<ul>
  <Todo
    id="#1580601600001"
  />
</ul>
`;

exports[`todo creation > 1. todo.onAdd > TodoToolbar (deep) 1`] = `
[
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
]
`;

exports[`todo creation > 1. todo.onAdd > TodoToolbar (shallow) 1`] = `
<React.Fragment>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  Hide completed
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>
</React.Fragment>
`;
